File: src/services/notes.ts
const API_BASE_URL = import.meta.env.VITE_API_URL || "http://localhost:8000";

export interface Note {
  _id: string;
  title: string;
  description?: string;
  cloudinary_url: string;
  cloudinary_public_id: string;
  thumbnail_url?: string;
  file_size?: number;
  uploaded_by: string;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

export interface NoteCreate {
  title: string;
  description?: string;
}

export interface NoteUpdate {
  title?: string;
  description?: string;
  is_active?: boolean;
}

export interface NoteListResponse {
  notes: Note[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface UploadNoteData {
  title: string;
  description?: string;
  file: File;
}

class NoteService {
  private getAuthHeaders(): HeadersInit {
    const token = localStorage.getItem('auth_access_token');
    return {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    };
  }

  private getUploadHeaders(): HeadersInit {
    const token = localStorage.getItem('auth_access_token');
    return {
      'Authorization': `Bearer ${token}`,
      // Don't set Content-Type for FormData, let browser set it
    };
  }

  async uploadNote(data: UploadNoteData): Promise<Note> {
    const formData = new FormData();
    formData.append('title', data.title);
    if (data.description) {
      formData.append('description', data.description);
    }
    formData.append('file', data.file);

    const response = await fetch(`${API_BASE_URL}/api/v1/notes/upload`, {
      method: 'POST',
      headers: this.getUploadHeaders(),
      body: formData,
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to upload note');
    }

    return response.json();
  }

  async getNotes(page: number = 1, pageSize: number = 10, search?: string): Promise<NoteListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString(),
    });

    if (search) {
      params.append('search', search);
    }

    const response = await fetch(`${API_BASE_URL}/api/v1/notes?${params.toString()}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch notes');
    }

    return response.json();
  }

  async getNote(noteId: string): Promise<Note> {
    const response = await fetch(`${API_BASE_URL}/api/v1/notes/${noteId}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch note');
    }

    return response.json();
  }

  async updateNote(noteId: string, data: NoteUpdate): Promise<Note> {
    const response = await fetch(`${API_BASE_URL}/api/v1/notes/${noteId}`, {
      method: 'PUT',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(data),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to update note');
    }

    return response.json();
  }

  async deleteNote(noteId: string): Promise<void> {
    const response = await fetch(`${API_BASE_URL}/api/v1/notes/${noteId}`, {
      method: 'DELETE',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to delete note');
    }
  }

  async getUserNotes(userId: string, page: number = 1, pageSize: number = 10): Promise<NoteListResponse> {
    const params = new URLSearchParams({
      page: page.toString(),
      page_size: pageSize.toString(),
    });

    const response = await fetch(`${API_BASE_URL}/api/v1/notes/user/${userId}?${params.toString()}`, {
      method: 'GET',
      headers: this.getAuthHeaders(),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.detail || 'Failed to fetch user notes');
    }

    return response.json();
  }

  // Helper function to generate thumbnail URL from PDF URL
  generateThumbnailUrl(pdfUrl: string): string {
    // Extract the public ID from the PDF URL
    // Example: https://res.cloudinary.com/dzacbdici/raw/upload/v1751991990/notes/Chemistry-book.pdf
    // Extract: v1751991990/notes/Chemistry-book
    const match = pdfUrl.match(/\/v(\d+)\/(.+)\.pdf$/);
    if (match) {
      const version = match[1];
      const publicId = match[2];
      // Return thumbnail URL: first page as JPG with dimensions
      return `https://res.cloudinary.com/dzacbdici/image/upload/c_fill,w_300,h_400,pg_1/v${version}/${publicId}.jpg`;
    }
    return ''; // Fallback if URL doesn't match expected pattern
  }
}

export const noteService = new NoteService();

File: backend/app/api/v1/endpoints/notes.py
from typing import Optional
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form, status
from fastapi.security import HTTPBearer
from app.models.note import NoteCreate, NoteUpdate, NoteResponse, NoteListResponse
from app.repositories.note_repository import NoteRepository
from app.config.cloudinary import CloudinaryService
from app.db.mongodb import get_database
from app.api.dependencies import get_current_user
from app.models.user import UserBase
import math

router = APIRouter()
security = HTTPBearer()


@router.post("/upload", response_model=NoteResponse)
async def upload_note(
    title: str = Form(..., min_length=1, max_length=200),
    description: Optional[str] = Form(None, max_length=1000),
    file: UploadFile = File(...),
    current_user: UserBase = Depends(get_current_user),
    db = Depends(get_database)
):
    """Upload a PDF note to Cloudinary and store metadata in database"""
    
    # Validate file type
    if not file.content_type == "application/pdf":
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Only PDF files are allowed"
        )
    
    # Validate file size (e.g., max 50MB)
    MAX_FILE_SIZE = 50 * 1024 * 1024  # 50MB
    file_content = await file.read()
    if len(file_content) > MAX_FILE_SIZE:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="File size too large. Maximum 50MB allowed."
        )
    
    try:
        # Upload to Cloudinary
        cloudinary_result = await CloudinaryService.upload_pdf(
            file_content, 
            file.filename.replace(".pdf", ""),
            folder="notes"
        )
        
        pdf_url, public_id, thumbnail_url = cloudinary_result
        
        if not pdf_url or not public_id:
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
                detail="Failed to upload file to Cloudinary"
            )
        
        # Create note data
        note_data = NoteCreate(title=title, description=description)
        
        # Save to database
        note_repo = NoteRepository(db)
        note = await note_repo.create_note(
            note_data=note_data,
            cloudinary_url=pdf_url,
            cloudinary_public_id=public_id,
            thumbnail_url=thumbnail_url,
            file_size=len(file_content),
            uploaded_by=str(current_user.id)
        )
        
        return NoteResponse(
            _id=str(note.id),
            title=note.title,
            description=note.description,
            cloudinary_url=note.cloudinary_url,
            cloudinary_public_id=note.cloudinary_public_id,
            thumbnail_url=note.thumbnail_url,
            file_size=note.file_size,
            uploaded_by=note.uploaded_by,
            created_at=note.created_at,
            updated_at=note.updated_at,
            is_active=note.is_active
        )
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to upload note: {str(e)}"
        )


@router.get("/", response_model=NoteListResponse)
async def get_notes(
    page: int = 1,
    page_size: int = 10,
    search: Optional[str] = None,
    db = Depends(get_database)
):
    """Get paginated list of notes"""
    
    if page < 1:
        page = 1
    if page_size < 1 or page_size > 100:
        page_size = 10
    
    skip = (page - 1) * page_size
    
    note_repo = NoteRepository(db)
    
    # Get notes and total count
    notes = await note_repo.get_notes(skip=skip, limit=page_size, search=search)
    total = await note_repo.count_notes(search=search)
    total_pages = math.ceil(total / page_size)
    
    note_responses = [
        NoteResponse(
            _id=str(note.id),
            title=note.title,
            description=note.description,
            cloudinary_url=note.cloudinary_url,
            cloudinary_public_id=note.cloudinary_public_id,
            thumbnail_url=note.thumbnail_url,
            file_size=note.file_size,
            uploaded_by=note.uploaded_by,
            created_at=note.created_at,
            updated_at=note.updated_at,
            is_active=note.is_active
        ) for note in notes
    ]
    
    return NoteListResponse(
        notes=note_responses,
        total=total,
        page=page,
        page_size=page_size,
        total_pages=total_pages
    )


@router.get("/{note_id}", response_model=NoteResponse)
async def get_note(
    note_id: str,
    db = Depends(get_database)
):
    """Get a specific note by ID"""
    
    note_repo = NoteRepository(db)
    note = await note_repo.get_note_by_id(note_id)
    
    if not note:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Note not found"
        )
    
    return NoteResponse(
        _id=str(note.id),
        title=note.title,
        description=note.description,
        cloudinary_url=note.cloudinary_url,
        cloudinary_public_id=note.cloudinary_public_id,
        thumbnail_url=note.thumbnail_url,
        file_size=note.file_size,
        uploaded_by=note.uploaded_by,
        created_at=note.created_at,
        updated_at=note.updated_at,
        is_active=note.is_active
    )


@router.put("/{note_id}", response_model=NoteResponse)
async def update_note(
    note_id: str,
    note_update: NoteUpdate,
    current_user: UserBase = Depends(get_current_user),
    db = Depends(get_database)
):
    """Update a note (title, description, or status)"""
    
    note_repo = NoteRepository(db)
    
    # Check if note exists and user has permission
    existing_note = await note_repo.get_note_by_id(note_id)
    if not existing_note:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Note not found"
        )
    
    # Allow update if user is the uploader or admin
    if existing_note.uploaded_by != str(current_user.id) and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to update this note"
        )
    
    updated_note = await note_repo.update_note(note_id, note_update)
    if not updated_note:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Failed to update note"
        )
    
    return NoteResponse(
        _id=str(updated_note.id),
        title=updated_note.title,
        description=updated_note.description,
        cloudinary_url=updated_note.cloudinary_url,
        cloudinary_public_id=updated_note.cloudinary_public_id,
        thumbnail_url=updated_note.thumbnail_url,
        file_size=updated_note.file_size,
        uploaded_by=updated_note.uploaded_by,
        created_at=updated_note.created_at,
        updated_at=updated_note.updated_at,
        is_active=updated_note.is_active
    )


@router.delete("/{note_id}")
async def delete_note(
    note_id: str,
    current_user: UserBase = Depends(get_current_user),
    db = Depends(get_database)
):
    """Delete a note (soft delete)"""
    
    note_repo = NoteRepository(db)
    
    # Check if note exists and user has permission
    existing_note = await note_repo.get_note_by_id(note_id)
    if not existing_note:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Note not found"
        )
    
    # Allow deletion if user is the uploader or admin
    if existing_note.uploaded_by != str(current_user.id) and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to delete this note"
        )
    
    success = await note_repo.delete_note(note_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Failed to delete note"
        )
    
    return {"message": "Note deleted successfully"}


@router.get("/user/{user_id}", response_model=NoteListResponse)
async def get_user_notes(
    user_id: str,
    page: int = 1,
    page_size: int = 10,
    current_user: UserBase = Depends(get_current_user),
    db = Depends(get_database)
):
    """Get notes uploaded by a specific user"""
    
    # Allow access if requesting own notes or admin
    if user_id != str(current_user.id) and not current_user.is_admin:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Not authorized to view these notes"
        )
    
    if page < 1:
        page = 1
    if page_size < 1 or page_size > 100:
        page_size = 10
    
    skip = (page - 1) * page_size
    
    note_repo = NoteRepository(db)
    notes = await note_repo.get_notes_by_user(user_id, skip=skip, limit=page_size)
    
    # Count total notes for the user
    total = await note_repo.count_notes()  # This would need to be modified to count by user
    total_pages = math.ceil(total / page_size)
    
    note_responses = [
        NoteResponse(
            _id=str(note.id),
            title=note.title,
            description=note.description,
            cloudinary_url=note.cloudinary_url,
            cloudinary_public_id=note.cloudinary_public_id,
            thumbnail_url=note.thumbnail_url,
            file_size=note.file_size,
            uploaded_by=note.uploaded_by,
            created_at=note.created_at,
            updated_at=note.updated_at,
            is_active=note.is_active
        ) for note in notes
    ]
    
    return NoteListResponse(
        notes=note_responses,
        total=len(note_responses),  # For now, return actual count
        page=page,
        page_size=page_size,
        total_pages=1  # For simplicity
    )
